tiledb/sm/buffer/buffer.h

```diff
@@ -418,6 +435,226 @@ class ConstBuffer : public BufferBase {
   }
 };
 
+/**
+ * Manages a byte buffer that is used for capnp (de)serialization.
+ *
+ * The buffer can be either owned by this class (happens in cases of
+ * serialization) or not (typically user-managed, happens in cases of
+ * deserialization).
+ *
+ * The buffer's content can be read from as a span of bytes. It cannot be
+ * modified, only replaced.
+ */
+class SerializationBuffer {
+ private:
+  /**
+   * Manages the memory of the buffer, if it is owned by this class.
+   *
+   * In the case of non-owned buffers, this vector is empty. The vector is not
+   * wrapped in an std::optional to avoid losing the allocator when reassigning
+   * between owned and non-owned buffers.
+   *
+   * @invariant Unless buffer_owner_ gets cleared to assign a non-owned buffer,
+   * after updating the content of buffer_owner_, buffer_ must be set to the
+   * memory pointed by buffer_owner_.
+   */
+  tdb::pmr::vector<char> buffer_owner_;
```

Comment:
`vector` unfortunately zero-initializes its elements which affects performance. I'm not sure if it does it when you call `vector::assign` or `reserve` (or if we can rely on it), but the overload that takes the size definitely does it (it could be argued that this is only being used by `tiledb_buffer_list_flatten` which is not performant by definition).

Using `unique_ptr` and manually implementing the allocator pattern is likely going to become quite more work.